Add `tags-dropdown` feature

HardikModha opened the pull request:
1. Here is the list of URLs that can be used to test the functionality.

- https://github.com/facebook/react/releases (Releases page)
- https://github.com/facebook/react/releases/tag/v16.8.1 (Release page of a particular tag)
- https://github.com/sindresorhus/refined-github/releases (Releases page with no releases)

2. It Fixes #1895 and closes #1895.

Note: The feature uses the GitHub API to fetch the releases for the project. If the personal token is available then it uses api v4 (GraphQL) otherwise it uses the api v3 (REST).

Demo:

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/22439276/55278732-06dd0c00-5336-11e9-8f52-dad791ccfd11.gif)

HardikModha commented:
cc @bfred-it @sindresorhus Please have a look. :)

bfred-it commented:
Isn’t this already available in the Code tab? There’s a branch/tag selector

HardikModha commented:
@bfred-it Yes, It's available on the Code tab but I believe when we want to search for the release notes of any particular release or want to download the release assets (e.g. binaries/tar) it will come handy.

bfred-it commented:
Right but I mean you can already search it from the Code tab without adding 250 lines of code

bfred-it commented:
Ah right, that one opens the code while this one would open the release directly.

I'm just concerned about the size. I'd like to see the native selector added instead, the code is just:

```html
<details class="details-reset details-overlay select-menu branch-select-menu">
  <summary class="btn btn-sm select-menu-button css-truncate"
           data-hotkey="w"
           
           title="Switch branches or tags">
    <i>Branch:</i>
    <span class="css-truncate-target">master</span>
  </summary>

  <details-menu class="select-menu-modal position-absolute" style="z-index: 99;" src="/sindresorhus/refined-github/ref-list/master?source_action=disambiguate&amp;source_controller=files" preload>
    <include-fragment class="select-menu-loading-overlay anim-pulse">
    </include-fragment>
  </details-menu>
</details>
```

And once it loads it can be trimmed of the "Branches" part and redirect the links to the release page instead of `/tree` (via `onclick` handler)

This frees us from any API, adds full search and we're not limited to the last 100 items

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,267 @@
+		return releases;
+	} catch (error) {
+		return Promise.reject(error);
+	}
```
`async` functions don't need try/catch, they already return a rejected `Promise`

HardikModha replied:
Code removed.

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,267 @@
+	}
+};
+
+const fetchReleasesV3 = async (ownerName, repoName): Promise<Release[]> => {
```
Generally speaking, only v4 is used, we can't provide two options. v3 is only used for things that v3 doesn't support.

HardikModha replied:
Code removed.

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,267 @@
+			return {tag: item.tag_name, htmlPage: item.html_url};
+		});
+
+		return releases;
```
If you're not going to use the variable, just return on line 118

HardikModha replied:
Code removed.

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,267 @@
+		</div>
+	);
+
+	container.append(child);
```
If you're not going to use the variable elsewhere, just return create the JSX inside `.append`

HardikModha replied:
Done.

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,267 @@
+	const selectedTag = getSelectedTagFromUrl();
+
+	return releases.map(release => {
+		return (
```
Use 

```js
return releases.map(release => (
```
instead of
```js
return releases.map(release => {
	return (
```

HardikModha replied:
Code removed.

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,267 @@
+			.substr(index)
+			.split('/')
+			.pop() :
+		null;
```
This function can be replaced with a regex, approximately:

```js
const [tag] = location.href.match(/releases\/tag\/(^\/+)/) || [];
return tag;
```

HardikModha replied:
Done. Replaced with `location.href.match(/releases\/tag\/([^>/]*)/)`. Your suggestion was not working.

HardikModha commented:
Yeah, You got it right. I didn't know about this `include-fragment`. I'm giving it a try and also address the changes. Thanks for your quick response :)

bfred-it commented:
`include-fragment` has a `load` event, if I remember correctly. Check out `on-new-comments.ts` for the example.

HardikModha commented:
Yep, It has. I am also going through the documentation. https://github.com/github/include-fragment-element

HardikModha commented:
@bfred-it A quick question. Adding include-fragment thing will require us to add two npm packages. 

1. [include-fragment](https://github.com/github/include-fragment-element) it self and
2. [Webcomponents pollyfill](https://github.com/webcomponents/custom-elements) for custom elements. 

Are there any concerns?

notlmn commented:
`<include-fragment>` is a custom element and GitHub has all of its code loaded already in the browser, you don't have to load it again, you just use it.

On usage of `<include-fragment>`, refer to https://github.com/sindresorhus/refined-github/blob/0b6c7330c21f8570519143ad3d0decc328d4db47/source/features/show-recently-pushed-branches-on-more-pages.tsx#L28

HardikModha commented:
@notlmn Thanks for your response. I initially thought the same but when I tried, It failed with `Cannot read property 'get' of null at window.customElements.get`. So I tried with those packages installed and it worked. Maybe I need to look into it again.

HardikModha commented:
@notlmn Don't know what was the issue. This time it worked without those packages installed. 🤔

HardikModha commented:
@bfred-it I have made the suggested changes. Please take a look. :)

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,155 @@
+const getSelectedTagFromUrl = () => {
+	const result = location.href.match(/releases\/tag\/([^>/]*)/) || [];
+
+	return (result && result.length > 0) ? result[1] : null;
```
Please use the suggestion I gave before (restructuring directly on the `const` so you don’t have to check for the length.

HardikModha replied:
TIL, location.href.match(/(?<=\/releases\/tag\/)([^\/]+)/).

HardikModha commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,155 @@
+
+async function init() {
+	if (!isReleasesOrTags()) {
+		return false;
```
Removed.

HardikModha commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,155 @@
+		return false;
+	}
+
+	if (!isOnReleasePage()) {
```
You are right but here we are already checking the page by `features.isReleasesOrTags`. I have added this check for the cases when a release is not present. e.g. release page of [refined-github](https://github.com/sindresorhus/refined-github/releases). I think the name is little ambiguos. Changing it to `hasAnyReleases`. What do you think?

bfred-it replied:
Gotcha, if it’s just checking if any releases exist at all (so “the feature isn’t needed” rather than “we’re not on the right page”) then it’s fine here in init. 

You can probably also inline a `select.exists` that looks for the “there are no releases” element. I think GitHub calls them “blankslate”

HardikModha replied:
I think looking for `.release` and `.release-entry` will suffice. No need for an extra check of `.blankslate`.

bfred-it replied:
It's not an extra check, but a single check for one element. Also generally you can combine selectors like `select.exists('.release, .release-entry')`

HardikModha replied:
Okay. Suggested included. 😅

HardikModha commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,155 @@
+	adjustListStyles();
+	removeHeader();
+	removeTabs();
+	updateLinks();
```
> These are all single-use functions. You can inline them all and avoid checking the elements in each.

The elements are being checked are different in each. I would say, I still check whether the element exists or not and if doesn't exist, log the error and return. What do you think?

> Usually, if the element isn’t present, we need the error to show so we can update the selector. `If false return` makes all errors silent.

Yeah, I agree with you here. Silently returning won't be helpful. 

Regarding the separation of function, I thought it would be better to keep them separate based on their functionality. But as they are very small, I will inline them.

bfred-it replied:
Why do you need the checks? To avoid errors? We need those errors to see when a features breaks. 

Also why do you want to log the error? Letting it throw already appears in the console and requires extra code. 

Most of those functions can be replaced by something like:

```js 
select('…').remove()
```

and 5 lines of that is more readable than 5 functions in 5 places amounting to at least 4 times the amount of lines each. It has no benefit and adds overhead for the reader. 

If a line does something incomprehensible, a comment can explain more than a function name (but in this case none of the function names help either)

HardikModha replied:
Yeah, You are right. Not avoiding the errors anymore. Also inlines all functions.

HardikModha commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,155 @@
+
+		item.setAttribute('href', link.replace('/tree/', '/releases/tag/'));
+
+		if (item.innerText.trim() === selectedTag) {
```
Done. 

**TIL**: `textContent` over `innerText`.

bfred-it replied:
Yeah, I think some of these changes will soon be enforced by the l’Inter so they’ll be caught early

HardikModha commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,155 @@
+};
+
+function deinit() {
+	deregisterEventListener();
```
Removed deinit.

HardikModha replied:
Removed deinit.

HardikModha commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,155 @@
+					src={`/${ownerName}/${repoName}/ref-list/master?source_action=disambiguate&source_controller=files`}
+					preload
+					role="menu"
+					style="z-index: 99;"
```
Ooops..Glad it works the other way. Fixed.

HardikModha commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,155 @@
+	const links = select.all('.js-release-details > tab-container > .select-menu-list > div .select-menu-item');
+	const selectedTag = getSelectedTagFromUrl();
+
+	links.forEach(item => {
```
Done.

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,105 @@
+	for (const item of links) {
+		const link = item.getAttribute('href');
+
+		item.setAttribute('href', link.replace('/tree/', '/releases/tag/'));
```
This will fail on repos called “tree”. Usually I use something like

```is
const url = link.pathname.split('/');
url[3] = 'releases/tag'; // Replaces `tree`
item.pathname = url;
```

also URL props are more convenient than attributes

HardikModha replied:
Great catch...Thanks for the advice. Will keep it in mind. :)

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,105 @@
+
+async function init() {
+	if (!hasAnyReleases()) {
+		console.debug('The repo does not have any releases.');
```
This is temporary right? Either way it can be dropped, we don’t have special debugging for each feature.

HardikModha replied:
Yeah. Removed.

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,105 @@
+	const selectedTag = getSelectedTagFromUrl();
+
+	container.append(
+		<div class="refined-github-search-releases float-right d-flex flex-shrink-0 flex-items-center mb-3">
```
Our class name can start with `rgh-`

HardikModha replied:
Removed the class name from here as it was not being used.

HardikModha commented:
@bfred-it Thank you so much for the in-depth reviews. It helped me a lot to understand some concepts and improve my knowledge. I've pushed the latest changes and I think this should suffice. Please let me know if any more changes are needed on my side. :)

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,94 @@
+import {octoface} from '../libs/icons';
+
+async function init() {
+	if (!hasAnyReleases()) {
```
```suggestion
	if (select.exists('.blankslate')) {
```

HardikModha replied:
Done. 👍

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,94 @@
+	const {ownerName, repoName} = getOwnerAndRepo();
+	const selectedTag = getSelectedTagFromUrl();
+
+	container.append(
```
```suggestion
	select('.subnav').append(
```

HardikModha replied:
Done. 👍

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,94 @@
+};
+
+const getSelectedTagFromUrl = () => {
+	const [tag = null] = location.href.match(/(?<=\/releases\/tag\/)([^/]+)/) || [];
```
```suggestion
	const [tag] = location.href.match(/(?<=\/releases\/tag\/)([^/]+)/) || [undefined];
```

I don't know if this works, but it would be preferred if it does: https://github.com/sindresorhus/meta/issues/7

HardikModha replied:
It works. Suggestion implemented.

bfred-it replied:
Yeah I hate this specific nitpick of TypeScript, I wish they had a better suggestion in the error message.

HardikModha replied:
I wish! But anyways I love Typescript! 🤪

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,94 @@
+					role="menu"
+					style={{zIndex: 99}}
+				>
+					<include-fragment class="select-menu-loading-overlay anim-pulse">
```
```suggestion
					<include-fragment class="select-menu-loading-overlay anim-pulse" onload={onFragmentLoaded}>
```

and then you can:
1. drop `registerEventListener`
2. merge `addSelectMenu` into `init`, like:

```js
async function init() {
	if (select.exists('.blankslate')) {
		return false;
	}

	const {ownerName, repoName} = getOwnerAndRepo();

	select('.subnav').append(
```

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,94 @@
+			<details class="details-reset details-overlay select-menu branch-select-menu">
+				<summary class="btn btn-sm select-menu-button css-truncate" data-hotkey="w" title="Filter releases" aria-haspopup="menu">
+					<i>Filter:</i>&nbsp;
+					<span class="css-truncate-target">{selectedTag || 'Releases'}</span>&nbsp;
```
```suggestion
					<span class="css-truncate-target">{getCurrentTag() || 'Releases'}</span>&nbsp;
```

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,94 @@
+	);
+};
+
+const getSelectedTagFromUrl = () => {
```
```suggestion
const getCurrentTag = () => {
```

HardikModha replied:
Done. 👍

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,94 @@
+
+	select('.rgh-release-details > .select-menu-header').remove();
+
+	select('.rgh-release-details > tab-container > .select-menu-filters > .select-menu-tabs').remove();
```
Can all of these selectors be simplified? For example, any of these:

- drop `>` if not necessary, like `.rgh-release-details tab-container .select-menu-list`
- reduce parts, like `.rgh-release-details .select-menu-list`

HardikModha replied:
I agree we can remove `.rgh-release-details` from some of them and simplify them. I'll simply them wherever possible. But I had kept the parent selector as `.rgh-release-details` so that all selector become scoped and we don't remove anything else accidentally (now or in future).

HardikModha replied:
Or Do I use `select` with a base element? 

```
select.exists('.foo a[href=bar]', baseElement)
```

bfred-it replied:
> I agree we can remove `.rgh-release-details` from some of them

I didn't say that 😁

I think that can stay the same, it's what comes after that should be simplified

HardikModha replied:
> `reduce parts, like .rgh-release-details .select-menu-list`

Maybe I misunderstood the above statement then. 😅

bfred-it replied:
"like" meant "for example, it should look like this"

Sorry for the confusion

HardikModha replied:
No worries! :)

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,94 @@
+
+	select('.rgh-release-details > tab-container > .select-menu-list').remove();
+
+	select('.rgh-release-details > tab-container > .select-menu-list').removeAttribute('hidden');
```
I think this selector is wrong, the same element is being `remove`d

HardikModha replied:
My bad. It's ambiguous because the same selector is present on two elements. How about doing?

```
const menu = select.all('tab-container > .select-menu-list');

menu[0].remove();
menu[1].removeAttribute('hidden');
```

bfred-it replied:
Gotcha. Yeah I think that would be better, also because it becomes one `select`

However I'd use destructuring:

```js
////// use the right names, I don't know what these are
const [nameThis, nameThat] = select.all('tab-container > .select-menu-list');
nameThis.remove();
nameThat.removeAttribute('hidden');
```

HardikModha replied:
Cool. Implemented. 👍

bfred-it replied:
Ah, I see... maybe we can just use:

```js
 	select('.rgh-release-details tab-container').remove(); 😅
```

HardikModha replied:
Yes, You got it right. We cannot. Because doing so will remove the lists, too.

bfred-it commented on source/features/search-releases.tsx:
```diff
@@ -0,0 +1,94 @@
+
+features.add({
+	id: 'search-releases',
+	include: [features.isReleasesOrTags],
```
```suggestion
	include: [
		features.isReleasesOrTags
	],
```

Not better, but all features are being written this way for consistency

bfred-it commented on source/libs/icons.tsx:
```diff
@@ -52,6 +52,8 @@ export const gear = (): SVGElement => <svg aria-hidden="true" class="octicon oct
 
 export const foldDown = (): SVGElement => <svg aria-hidden="true" class="octicon octicon-fold-down" width="14" height="16" ><path fill-rule="evenodd" d="M4 11l3 3 3-3H8V5H6v6H4zm-4 0c0 .55.45 1 1 1h2.5l-1-1h-1l2-2H5V8H3.5l-2-2H5V5H1c-.55 0-1 .45-1 1l2.5 2.5L0 11zm10.5-2H9V8h1.5l2-2H9V5h4c.55 0 1 .45 1 1l-2.5 2.5L14 11c0 .55-.45 1-1 1h-2.5l1-1h1l-2-2z"/></svg>;
 
+export const octoface = (): SVGElement => <svg height="32" class="octicon octicon-octoface" viewBox="0 0 16 16" version="1.1" width="32" aria-hidden="true"><path fill-rule="evenodd" d="M14.7 5.34c.13-.32.55-1.59-.13-3.31 0 0-1.05-.33-3.44 1.3-1-.28-2.07-.32-3.13-.32s-2.13.04-3.13.32c-2.39-1.64-3.44-1.3-3.44-1.3-.68 1.72-.26 2.99-.13 3.31C.49 6.21 0 7.33 0 8.69 0 13.84 3.33 15 7.98 15S16 13.84 16  ...
```
```suggestion
export const octoface = (): SVGElement => <svg aria-hidden="true" class="octicon octicon-octoface" width="32" height="32"><path d="M14.7 5.34c.13-.32.55-1.59-.13-3.31 0 0-1.05-.33-3.44 1.3-1-.28-2.07-.32-3.13-.32s-2.13.04-3.13.32c-2.39-1.64-3.44-1.3-3.44-1.3-.68 1.72-.26 2.99-.13 3.31C.49 6.21 0 7.33 0 8.69 0 13.84 3.33 15 7.98 15S16 13.84 16 8.69c0-1.36-.49-2.48-1.3-3.35zM8 14.02c-3.3 0-5.98-.15-5.98-3.35 0-.76.38-1.48 1.02-2.07 1.07-.98 2.9-.46 4.96-.46 2.07 0 3.88-.52 4.96.46.65.59 1.02 1.3 1.02 2.07 0 3.19-2.68 3.35-5.98 3.35zM5.49 9.01c-.66 0-1.2.8-1.2 1.78s.54 1.79 1.2 1.79c.66 0 1.2-.8 1.2-1.79s-.54-1.78-1.2-1.78zm5.02 0c-.66 0-1.2.79-1.2 1.78s.54 1.79 1.2 1.79c.66 0 1.2-.8 1.2-1.79s-.53-1.78-1.2-1.78z" /></svg>;
```

bfred-it replied:
You can commit this directly from GitHub in 2 clicks (Add suggestion to batch)

HardikModha commented:
@bfred-it Hopefully this should be it. I've implemented all the suggested changes. Please let me know if any more changes are needed. 

PS: If anything happens to me, my last wish is to see this PR getting merged. 😂

bfred-it commented:
Example:

<img width="424" alt="smaller" src="https://user-images.githubusercontent.com/1402241/55493151-0496ec80-566b-11e9-8cd4-42f319bc81b6.png">

Tested on 

https://github.com/bfred-it/doma/releases
https://github.com/facebook/react/releases/tag/v16.6.0-alpha.8af6728

I dropped the text from the button because it's duplicate in this case. In the Code tab, the switcher acts as a title, as it's the only place where "Branch: master" is shown, for example. In our case if there's a tag selected, we already have the title in a big `h1` on the page

bfred-it commented:
Also if you're a big tags user, this feature would be pretty useful: https://github.com/sindresorhus/refined-github/issues/966

HardikModha commented:
Changes look pretty good. I completely missed the drafting new release page. Thanks for making it perfect and putting efforts in reviewing the PR. :)

HardikModha commented:
@bfred-it I would like to contribute to #966. :)

HardikModha commented:
@bfred-it Can we merge this? or it will go with #966?

bfred-it commented:
Let’s wait for Sindre

sindresorhus commented:
@HardikModha Can you fix the TS issues in https://travis-ci.org/sindresorhus/refined-github/builds/521256206?utm_source=github_status&utm_medium=notification ? We enabled [TS strict mode](https://github.com/sindresorhus/refined-github/commit/7b3b275ad9a1b1409c56310c0b8ab286bef1f75f) recently in master.

sindresorhus commented on source/features/tags-dropdown.tsx:
```diff
@@ -0,0 +1,61 @@
+
+	// Change links (which point to the content of each tag) to open the tag page instead
+	for (const item of select.all<HTMLAnchorElement>('.rgh-tags-dropdown [href*="/tree/"]')) {
+		const arr = item.pathname.split('/');
```
`arr` needs a more descriptive name.

HardikModha replied:
How about `urlParts` or `splitPathName`?

bfred-it replied:
I used urlParts in other parts of the code

sindresorhus replied:
I prefer `pathnameParts`

sindresorhus commented on source/features/tags-dropdown.tsx:
```diff
@@ -0,0 +1,61 @@
+	select('.rgh-tags-dropdown .select-menu-tab:last-child button').click();
+
+	// Change links (which point to the content of each tag) to open the tag page instead
+	for (const item of select.all<HTMLAnchorElement>('.rgh-tags-dropdown [href*="/tree/"]')) {
```
```suggestion
	for (const anchorElement of select.all<HTMLAnchorElement>('.rgh-tags-dropdown [href*="/tree/"]')) {
```

sindresorhus commented on source/features/tags-dropdown.tsx:
```diff
@@ -0,0 +1,61 @@
+}
+
+function getCurrentTag() {
+	const [tag] = location.href.match(/(?<=\/releases\/tag\/)([^/]+)/) || [undefined];
```
Better to use `location.pathname`.

sindresorhus commented on source/features/tags-dropdown.tsx:
```diff
@@ -0,0 +1,61 @@
+}
+
+function getCurrentTag() {
+	const [tag] = location.href.match(/(?<=\/releases\/tag\/)([^/]+)/) || [undefined];
```
```suggestion
	const [tag] = location.href.match(/(?<=\/releases\/tag\/)([^/]+)/) || [];
```

HardikModha replied:
Removing `undefined` throws a type error 
```
Initializer provides no value for this binding element and the binding element has no default value
```

HardikModha replied:
Method removed as passing tag as a parameter was not making any difference.

bfred-it replied:
@sindresorhus we need to decide which one to use:

```js
const [a] = b.match(c) || [undefined];
const [a = undefined] = b.match(c) || [
];
```

I always suggest the former because it feels less noisy than the latter, but as seen in another PR, if the regex captures groups then it’s easier to use `[, a = undefined]`

sindresorhus commented on source/features/tags-dropdown.tsx:
```diff
@@ -0,0 +1,61 @@
+import React from 'dom-chef';
```
Should have a comment at the top here that succinctly describes the feature.

HardikModha replied:
Sure.

sindresorhus commented on source/libs/icons.tsx:
```diff
@@ -52,6 +52,9 @@ export const gear = (): JSX.Element => <svg aria-hidden="true" className="octico
 
 export const foldDown = (): JSX.Element => <svg aria-hidden="true" className="octicon octicon-fold-down" width="14" height="16" ><path fill-rule="evenodd" d="M4 11l3 3 3-3H8V5H6v6H4zm-4 0c0 .55.45 1 1 1h2.5l-1-1h-1l2-2H5V8H3.5l-2-2H5V5H1c-.55 0-1 .45-1 1l2.5 2.5L0 11zm10.5-2H9V8h1.5l2-2H9V5h4c.55 0 1 .45 1 1l-2.5 2.5L14 11c0 .55-.45 1-1 1h-2.5l1-1h1l-2-2z"/></svg>;
 
+// The viewBox is actually needed here
```
@bfred-it Why are we not using `viewBox` on all of the icon? It let's use resize any icon while preserving aspect ratio.

bfred-it replied:
Because a year ago we realized it wasn’t needed on any but a few, so we dropped the “cruft”

HardikModha commented:
@sindresorhus First of all, Thank you for your time to carry out a detailed review. I have made the suggested changes. Please have a look.

sindresorhus commented:
You seem to have overwritten my changes in https://github.com/sindresorhus/refined-github/pull/1896/commits/f5572b01031fdd5b7a8fff76cb6caf9ceb3a1b70.

sindresorhus commented:
The feature seems to work well.

sindresorhus commented:
@HardikModha Can you mention it in the readme and link to a screenshot of the feature (should be retina, if you don't have retina screen, then 200% zoom).

HardikModha commented:
> You seem to have overwritten my changes in [f5572b0](https://github.com/sindresorhus/refined-github/commit/f5572b01031fdd5b7a8fff76cb6caf9ceb3a1b70).

Sorry, It was not intentional. I've added your changes back. It should be good now. I've also updated the readme. Can you please take a look?

sindresorhus commented:
@HardikModha Same questions as https://github.com/sindresorhus/refined-github/pull/1904#issuecomment-484641357

HardikModha commented:
@sindresorhus It's a yes from me! :)

jerone commented:
The tags dropdown seems to align weirdly...

![image](https://user-images.githubusercontent.com/55841/56472481-265fe400-645f-11e9-8da0-d982bf364481.png)

bfred-it commented:
I opened a new issue

bfred-it commented:
@HardikModha if not too difficult, could you make the content load on demand?

For example in the sidebar here, when you click on Projects, the content will start loading. I think this can be achieved with some HTML. Here's the source of the Projects dropdown:

```html
<details class="details-reset details-overlay select-menu js-select-menu js-dropdown-details hx_rsm js-load-contents "
	data-contents-url="/sindresorhus/refined-github/issues/1896/show_partial?partial=issues%2Fsidebar%2Fprojects_menu_content"
	id="projects-select-menu" data-multiple>

	<summary class="text-bold discussion-sidebar-heading discussion-sidebar-toggle hx_rsm-trigger" aria-haspopup="menu"
		data-hotkey="p">
		<svg class="octicon octicon-gear" viewBox="0 0 14 16" version="1.1" width="14" height="16" aria-hidden="true">
			<path fill-rule="evenodd"
				d="M14 8.77v-1.6l-1.94-.64-.45-1.09.88-1.84-1.13-1.13-1.81.91-1.09-.45-.69-1.92h-1.6l-.63 1.94-1.11.45-1.84-.88-1.13 1.13.91 1.81-.45 1.09L0 7.23v1.59l1.94.64.45 1.09-.88 1.84 1.13 1.13 1.81-.91 1.09.45.69 1.92h1.59l.63-1.94 1.11-.45 1.84.88 1.13-1.13-.92-1.81.47-1.09L14 8.75v.02zM7 11c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z" />
		</svg>
		Projects
	</summary>

	<div class="select-menu-modal js-navigation-container position-absolute right-0 hx_rsm-modal"
		style="z-index: 99; overflow: visible;">
		<div class="select-menu-header">
			<span class="select-menu-title">Projects</span>
			<button class="hx_rsm-close-button d-none btn-link close-button" type="button"
				data-toggle-for="projects-select-menu"><svg aria-label="Close menu" class="octicon octicon-x"
					viewBox="0 0 12 16" version="1.1" width="12" height="16" role="img">
					<path fill-rule="evenodd"
						d="M7.48 8l3.75 3.75-1.48 1.48L6 9.48l-3.75 3.75-1.48-1.48L4.52 8 .77 4.25l1.48-1.48L6 6.52l3.75-3.75 1.48 1.48L7.48 8z" />
				</svg></button>
		</div>

		<div class="js-select-menu-deferred-content hx_rsm-content "
			data-url="/sindresorhus/refined-github/issues/1896/show_partial?partial=issues%2Fsidebar%2Fprojects_menu_content">

		</div>
		<div class="select-menu-loading-overlay anim-pulse hx_rsm-loading">
			<svg height="32" class="octicon octicon-octoface" viewBox="0 0 16 16" version="1.1" width="32" aria-hidden="true">
				<path fill-rule="evenodd"
					d="M14.7 5.34c.13-.32.55-1.59-.13-3.31 0 0-1.05-.33-3.44 1.3-1-.28-2.07-.32-3.13-.32s-2.13.04-3.13.32c-2.39-1.64-3.44-1.3-3.44-1.3-.68 1.72-.26 2.99-.13 3.31C.49 6.21 0 7.33 0 8.69 0 13.84 3.33 15 7.98 15S16 13.84 16 8.69c0-1.36-.49-2.48-1.3-3.35zM8 14.02c-3.3 0-5.98-.15-5.98-3.35 0-.76.38-1.48 1.02-2.07 1.07-.98 2.9-.46 4.96-.46 2.07 0 3.88-.52 4.96.46.65.59 1.02 1.3 1.02 2.07 0 3.19-2.68 3.35-5.98 3.35zM5.49 9.01c-.66 0-1.2.8-1.2 1.78s.54 1.79 1.2 1.79c.66 0 1.2-.8 1.2-1.79s-.54-1.78-1.2-1.78zm5.02 0c-.66 0-1.2.79-1.2 1.78s.54 1.79 1.2 1.79c.66 0 1.2-.8 1.2-1.79s-.53-1.78-1.2-1.78z" />
			</svg>
		</div>
	</div>
</details>
```
